src/Interpreters/HashJoin.cpp

```diff
@@ -808,17 +935,145 @@ class AddedColumns
         type_name.emplace_back(src_column.type, src_column.name, qualified_name);
     }
 };
+using AddedColumnsV = std::vector<std::unique_ptr<AddedColumns>>;
+
+namespace
+{
+template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS>
+struct JoinFeatures
+{
+    static constexpr bool is_any_join = STRICTNESS == ASTTableJoin::Strictness::Any;
+    static constexpr bool is_all_join = STRICTNESS == ASTTableJoin::Strictness::All;
+    static constexpr bool is_asof_join = STRICTNESS == ASTTableJoin::Strictness::Asof;
+    static constexpr bool is_semi_join = STRICTNESS == ASTTableJoin::Strictness::Semi;
+    static constexpr bool is_anti_join = STRICTNESS == ASTTableJoin::Strictness::Anti;
+
+    static constexpr bool left = KIND == ASTTableJoin::Kind::Left;
+    static constexpr bool right = KIND == ASTTableJoin::Kind::Right;
+    static constexpr bool inner = KIND == ASTTableJoin::Kind::Inner;
+    static constexpr bool full = KIND == ASTTableJoin::Kind::Full;
+
+    static constexpr bool need_replication = is_all_join || (is_any_join && right) || (is_semi_join && right);
+    static constexpr bool need_filter = !need_replication && (inner || right || (is_semi_join && left) || (is_anti_join && left));
+    static constexpr bool add_missing = (left || full) && !is_semi_join;
+
+    static constexpr bool need_flags = MapGetter<KIND, STRICTNESS>::flagged;
+
+};
+
+template <bool multiple_disjuncts>
+class KnownRowsHolder;
 
-template <typename Map, bool add_missing>
-void addFoundRowAll(const typename Map::mapped_type & mapped, AddedColumns & added, IColumn::Offset & current_offset)
+// TODO: helper to clean, instead of recreating
+
+template<>
+class KnownRowsHolder<true>
+{
+public:
+    using Type = std::pair<const Block*, DB::RowRef::SizeT>;
+
+private:
+    static const size_t MAX_LINEAR = 16;
```

Comment:
Not at all.
We are memorizing rows from the left (I think) side of join to avoid duplicates. If we have "true OR true OR true" for a particular pair of rows, we do not want to have this pair three times.
"16" is a threshold we switch from array to set.
There is a test that covers it.

Reply:
Do we actually need this dynamic switching? Does it really affects performance?

Reply:
Frankly speaking, I do not know. 
If all conditions (disjuncts) produce multitude of matches for some rows, using set makes a lot of sense. Imagine, we've got million of matches from the first disjunct anf same number from the second one. 
If number of matches is always small, set will be probably slower than array.
Complexity is O(x^2) for array and O(x*log(x)) for set.
I believe that it is possible to make up examples where either choice would be suboptimal, that's why I've introduced this dynamic switch.